server/src/main/java/org/elasticsearch/action/DocWriteRequest.java

```diff
@@ -60,6 +60,15 @@
      */
     String type();
 
+    /**
+     * Set the default type supplied to a bulk
+     * request if this individual request's type is null
+     * or empty
+     * @return the Request
+     */
+    T defaultTypeIfNull(String defaultType);
```

Comment:
Can you briefly explain why this is added? Only to improve my understanding of the PR.

Reply:
Bulk requests can specify a default index and type which is automatically filled-in on any contained Index/Update/DeleteRequests that are missing these properties. A form of inheritance.

This is the method where any blank choice of type is overridden with the BulkRequest's choice of type. This code became necessary after a different PR switched the Index/Update/DeleteRequests to default to `_doc` instead of null - all the bulk type inheritance tests failed because we couldn't differentiate between objects where the user had explicitly defined a choice of `_doc` and those where the user had not made an explicit choice. Here we have to revert back to holding a null for this setting by default.

Reply:
Thanks for the explanation, think I got it.